qiskit/circuit/quantumcircuit.py

```diff
@@ -239,6 +244,24 @@ def data(self, data_input):
         for inst, qargs, cargs in data_input:
             self.append(inst, qargs, cargs)
 
+    @property
+    def metadata(self):
+        """The user provided metadata associated with the circuit
+
+        The metadata for the circuit is a user provided ``dict`` of metadata
+        for the circuit. It will not be used to influence the execution or
+        operation of the circuit, but it is expected to be passed betweeen
+        all transforms of the circuit (ie transpilation) and that providers will
+        associate any circuit metadata with the results it returns from
+        execution of that circuit.
+        """
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, metadata):
```

Comment:
Maybe we should do a type check here that metadata is a dict

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5270/commits/26036dc51ee57c5d4a7dd6bd866d73bf07233cb7